ocaml/testsuite/tests/warnings/w53.ml

```diff
@@ -473,3 +473,61 @@ module TestLayoutPolyStruct = struct
   external y : (int64 [@layout_poly]) -> (int64 [@layout_poly]) = "%identity" (* rejected *)
   external z : ('a : any). 'a -> 'a = "%identity" [@@layout_poly] (* accepted *)
 end
+
+module type TestZeroAllocSig = sig
+  type 'a t1 = 'a [@@zero_alloc] (* rejected *)
+  type s1 = Foo1 [@zero_alloc] (* rejected *)
+  val x : int [@@zero_alloc] (* rejected *)
+  val f : int -> int [@@zero_alloc] (* rejected *)
+
+  external y : (int [@zero_alloc]) -> (int [@zero_alloc]) = "x" (* rejected *)
+  external z : int -> int = "x" "y" [@@zero_alloc] (* rejected *)
+  external[@zero_alloc] q : int -> int = "x" "y" (* rejected *)
+
+  class[@zero_alloc] c : (* rejected *)
+    object
+      val[@zero_alloc] foo : int * int (* rejected *)
+      val[@zero_alloc] bar : int -> int (* rejected *)
+      method[@zero_alloc] baz : int * int (* rejected *)
+      method[@zero_alloc] boz : int -> int (* rejected *)
+    end
+end
+
+module TestZeroAllocStruct = struct
+  type 'a t1 = 'a [@@zero_alloc] (* rejected *)
+  type s1 = Foo1 [@zero_alloc] (* rejected *)
+  let x : int = 42 [@@zero_alloc] (* rejected *)
+
+  let[@zero_alloc] w = 42 (* rejected *)
+
+  let[@zero_alloc] f x = x (* accepted *)
+
+  external y : (int [@zero_alloc]) -> (int [@zero_alloc]) = "x" (* rejected *)
+  external z : int -> int = "x" "y" [@@zero_alloc] (* rejected *)
+  external[@zero_alloc] q : int -> int = "x" "y" (* rejected *)
+
+  class[@zero_alloc] foo _y = (* rejected *)
+    let[@inline never][@zero_alloc] f x = (x, x) in (* accepted *)
+    (fun[@zero_alloc] z -> (* rejected *)
+    object
+      val[@zero_alloc] bar = (4, 5) (* rejected *)
+
+      method[@zero_alloc] baz x = (f (z+10), x+1) (* rejected *)
+    end)
+
```

Comment:
I'm not excited about minting 2 more files for misplaced attribute tests. (There's already 2, for ppxes vs no-ppxes.)
* Eventually, could `[@@@zero_alloc all]` be scoped to the submodule it appears in, like `[@@@ocaml.warning ...]` attributes?
* For the time being, maybe we could just put the `[@@@zero_alloc all]` test last.

Reply:
> I'm not excited about minting 2 more files for misplaced attribute tests. (There's already 2, for ppxes vs no-ppxes.)
> 
> * Eventually, could `[@@@zero_alloc all]` be scoped to the submodule it appears in, like `[@@@ocaml.warning ...]` attributes?

yes, we had at go at it a while ago https://github.com/ocaml-flambda/flambda-backend/pull/1559 , it's still something the user want, and it maybe be easier to implement now that the parsing of these attributes on functions is earlier.

> * For the time being, maybe we could just put the `[@@@zero_alloc all]` test last.

It doesn't work like this, unfortunately, in the current `main`.  No matter where it is in the file, its effect is for the entire file (because its effect is applied later, after parsing).

Reply:
Ah, I see. In that case I'll retire my objection about separate files.

Reply:
I am a bit confused by the above conversation.  `zero_alloc all` changes the meaning of `zero_alloc` attributes, but I do not think it changes whether or not we expect w53 to fire for them.

I added a test that demonstrates that.  It is just a copy of the existing w53 tests for zero_alloc with `[@@@zero_alloc all]` added to the top, showing that the same things do/don't result in w53.

Along the way, I figured out how to share reference files for the normal and marshalled versions of these tests, so you'll see I also deleted `w53_marshalled.compilers.reference`.  That test now uses the same reference file as the normal `w53` test.  And the normal/marshalled versions of this new test share one new reference file.

Reply:
I didn’t understand that `all` didn’t affect w53 for other zero-alloc attributes. (You could imagine it doing so.) In any case, your approach sounds good, and thanks also for getting rid of the extra reference file.

Reply:
The new tests look fine to me.